### PR TITLE
[Feature]Add coco, lvis and voc dataset download script based on pr #6715

### PR DESCRIPTION
**Pull request by Czm369**
## Motivation

Add a script to download coco2017, lvis, voc2007 dataset based on https://github.com/open-mmlab/mmdetection/pull/6715

## Modification

Add a python file in tools/misc

## BC-breaking (Optional)

This script do not link any other files

## Use cases (Optional)

You can use command "python tools/download.py " to download coco2017 dataset.
Linux for using command "python tools/misc/download.py
You can alse use flag --unzip to unzip the zipfiles
such as python tools/misc/download.py  --unzip

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tools/misc/download_dataset.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import argparse
+import os
+import sys
+from itertools import repeat
+from multiprocessing.pool import ThreadPool
+from pathlib import Path
```
Comment (ZwwWayne) [inline tools/misc/download_dataset.py:6]: we usually use os.path rather than pathlib

Comment (Czm369) [reply tools/misc/download_dataset.py:6]: In my view, the author want to call the function path.mkdir(parents=True, exist_ok=True) to make a directory more securely, which may be convenient to download files in multiple threads.

File: tools/misc/download_dataset.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        '--unzip',
+        action='store_true',
+        help='whether unzip dataset or not, zipped files will be saved')
+    parser.add_argument(
+        '--delete',
+        action='store_true',
+        help='delete the download zipped files')
+    parser.add_argument(
+        '--threads', type=int, help='number of threading', default=4)
+    args = parser.parse_args()
```
Comment (RangiLyu) [inline tools/misc/download_dataset.py:99]: In mmdet we usually define a `parse_args():` function and call it in `main()`. You can refer to other tools.

**Comment by RangiLyu** [comment]
Add documentation of this tool in docs/en/1_exist_data_model.md and docs/en/useful_tools.md.

File: docs/en/1_exist_data_model.md [hunk 3]
```diff
@@ -173,6 +173,7 @@ This section will show how to test existing models on supported datasets.
 Public datasets like [Pascal VOC](http://host.robots.ox.ac.uk/pascal/VOC/index.html) or mirror and [COCO](https://cocodataset.org/#download) are available from official websites or mirrors. Note: In the detection task, Pascal VOC 2012 is an extension of Pascal VOC 2007 without overlap, and we usually use them together.
 It is recommended to download and extract the dataset somewhere outside the project directory and symlink the dataset root to `$MMDETECTION/data` as below.
 If your folder structure is different, you may need to change the corresponding paths in config files.
+We provide a script to download datasets such as COCO , you can run `python tools/misc/download_dataset.py --dataset-name coco2017` to download COCO dataset.
```
Comment (RangiLyu) [inline docs/en/1_exist_data_model.md:176]:
```suggestion

We provide a script to download datasets such as COCO , you can run `python tools/misc/download_dataset.py --dataset-name coco2017` to download COCO dataset. 

For more usage please refer to [dataset-download](https://github.com/open-mmlab/mmdetection/tree/master/docs/en/useful_tools.md#dataset-download)
```

File: tools/misc/download_dataset.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            print('Unzipping {}'.format(f.name))
+            if f.suffix == '.zip':
+                ZipFile(f).extractall(path=dir)
+            elif f.suffix == '.gz':
+                os.system(f'tar xfz {f} --directory {f.parent}')
+            if delete:
+                f.unlink()
+
+    if not url:
+        print('Only support coco now!')
```
Comment (RangiLyu) [inline tools/misc/download_dataset.py:58]: more datasets are supported now

File: tools/misc/download_dataset.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            download_one(u, dir)
+
+
+def main():
+    args = parse_args()
+    path = Path(args.save_dir)
+    if not path.exists():
+        path.mkdir(parents=True, exist_ok=True)
+    data2url = dict(
+        coco2017=[
```
Comment (hhaAndroid) [inline tools/misc/download_dataset.py:78]: TODO

File: tools/misc/download_dataset.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            elif f.suffix == '.gz':
+                os.system(f'tar xfz {f} --directory {f.parent}')
+            if delete:
+                f.unlink()
+
+    if not url:
+        print('Only support coco, voc, and lvis now!')
+        return
+    dir = Path(dir)
+    dir.mkdir(parents=True, exist_ok=True)
```
Comment (hhaAndroid) [inline tools/misc/download_dataset.py:61]: duplicate

File: tools/misc/download_dataset.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            else:
+                torch.hub.download_url_to_file(url, f, progress=True)
+        if unzip and f.suffix in ('.zip', '.gz'):
+            print('Unzipping {}'.format(f.name))
+            if f.suffix == '.zip':
+                ZipFile(f).extractall(path=dir)
+            elif f.suffix == '.gz':
+                os.system(f'tar xfz {f} --directory {f.parent}')
+            if delete:
+                f.unlink()
```
Comment (hhaAndroid) [inline tools/misc/download_dataset.py:55]: Deleting files should prompt the user

File: tools/misc/download_dataset.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            if curl:
+                os.system('curl -L {} -o {} --retry 9 -C -'.format(url, f))
+            else:
+                torch.hub.download_url_to_file(url, f, progress=True)
+        if unzip and f.suffix in ('.zip', '.gz'):
+            print('Unzipping {}'.format(f.name))
+            if f.suffix == '.zip':
+                ZipFile(f).extractall(path=dir)
+            elif f.suffix == '.gz':
+                os.system(f'tar xfz {f} --directory {f.parent}')
```
Comment (hhaAndroid) [inline tools/misc/download_dataset.py:53]: Is window feasible

File: tools/misc/download_dataset.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            if delete:
+                f.unlink()
+
+    if not url:
+        print('Only support coco, voc, and lvis now!')
```
Comment (hhaAndroid) [inline tools/misc/download_dataset.py:58]: The location of the prompt information is unreasonable

**Review by hhaAndroid — APPROVED** [review]

File: docs/en/useful_tools.md [hunk 10]
```diff
@@ -377,6 +377,14 @@ python tools/dataset_converters/cityscapes.py ${CITYSCAPES_PATH} [-h] [--img-dir
 python tools/dataset_converters/pascal_voc.py ${DEVKIT_PATH} [-h] [-o ${OUT_DIR}]
 ```
 
+## Dataset Download
+
+`tools/misc/download_dataset.py` supports downloading datasets such as COCO, VOC, and LVIS.
+```shell
```
Comment (ZwwWayne) [inline docs/en/useful_tools.md:383]: empty line above code block

**Pull request merged**